### ISSUE TITLE
a priori knowledge of database address

### ISSUE THREAD

**Issue by koh-osug** [orbitdb/orbitdb#1141]
My goal is to deploy a distributed P2P database among a couple of independent clients. Now all the clients have no knowledge of each other but shall use the same database with the same data. In the tutorials I have only found that one node is creating a database and all other nodes have to be aware of this database. Is there an option to have a fixed in advance generated address so that no special master node must be used to create the initial database? Or could this database be created on a PC and then deployed as initial storage to all the nodes?

**Comment by haydenyoung** [comment]
Yes, you can. An orbitdb address is just an IPFS cid hashed using base58 encoding. Theoretically, you could [generate the address yourself using Block.encode, passing in the details of the database](https://github.com/orbitdb/orbitdb/blob/c72b8fe2432398affda14132b78dfdeebf12d274/src/manifest-store.js#L44). However, generating the same address on a remote peer would require the same identity information. Therefore, care must be taken when using the same private keys across multiple systems.

**Comment by haydenyoung** [comment]
However, I should add that you will need to create the db by name at some stage. You cannot open a database by its address if the database does not exist.

**Comment by koh-osug** [comment]
The background of my question is that I want to share the same data between several nodes. And since all nodes are equal I have a chicken egg problem. Which peer will create the DB? And how are the others getting the DB address?

**Comment by haydenyoung** [comment]
> Which peer will create the DB?

As outlined above, you can create the database on each peer using the same name and identity but this would require sharing the same private key among several nodes. If this is something you can do securely, you could create a database on each peer without sharing the address.

> And how are the others getting the DB address?

How you share the database address is determined by your software's requirements. It could be as simple as posting the db address somewhere. A more automated solution might use pubsub to publish an address to subscribed peers. Alternatively, as I outlined above, you can determine the database address using `Block.encode` but make sure you are passing the same name, metadata, identity, etc.

**Comment by koh-osug** [comment]
I think the approach with libp2p is OK and I can publish the different addresses and keep different identities. Are the other nodes automatically updated or do I always have do use an `db.events.on('update'` listener and call `all`?
Also is a call to `all` dangerous? I have seen this in the listener examples. The documentation says that it list all documents, in case of 10000 and more this seems to be very expensive? Is there an alternative way to wait until all data is available?

**Comment by haydenyoung** [comment]
If the nodes are connected to each other they will sync (I.e. eventual consistency).

You do not need to use `all` to retrieve all records. In fact, in the latest version of OrbitDB (I.e. 1.0.2), OrbitDB peers will sync all records, there is no longer the concept of "lazy loading" (previous versions of OrbitDB v1 synced only the database "heads", heads being the latest records in a database across all orbitdb peers). Therefore, joining a new node and syncing it will be a costly exercise at the start but once sync-ing is complete, subsequent updates should be more efficient.

Latest documentation should reflect the changes to record sync-ing. If not, updates will be made in the next version of OrbitDB.

**Comment by koh-osug** [comment]
This sounds great. However if the sync is happening automatically how can I get the `update` event? What I notice here is that the db is started with `syncAutomatically` on and seems to fetch already some items and my `update` handler sees only the last added entry.

~~~javascript
const db2 = await orbitdb2.open(argv.dbAddress)

db2.events.on('join', async (peerId, heads) => {
  console.log("peer joining: " + peerId)
})

let updateCount=0
db2.events.on('update', async (entry) => {
  console.log(`new entry ${updateCount++}:\n`, entry)
})
~~~

Do I have to start with  `syncAutomatically`  somehow add my handler first and then turn on ` syncAutomatically` ?

**Comment by haydenyoung** [comment]
I was unable to replicate your issue. Implementing the same code results in update being fired 4 times for the 4 records added to my replicated peer.

Perhaps if you can expand on what you are using the `update` event for.

**Comment by koh-osug** [comment]
What I do is to start `node orbit1.js` on the first node. 10 items are created. I see
> Swarm listening on /ip4/127.0.0.1/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/192.168.178.151/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/172.17.0.1/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/172.25.0.1/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/10.63.170.1/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/192.168.1.101/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/192.168.1.102/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
Swarm listening on /ip4/192.168.224.1/tcp/46433/p2p/12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
db1 address:  /orbitdb/zdpuArrSWat42iJuw8LTqHQFQw3GvJaCGqMmikB48QNTwDTPQ
putting data 0
putting data 1
putting data 2
putting data 3
putting data 4
putting data 5
putting data 6
putting data 7
putting data 8
putting data 9
Script is running. Press CTRL+C to terminate.

Then I start on a second computer `orbit1.js` with `node orbit2.js --dbAddress=/orbitdb/zdpuArrSWat42iJuw8LTqHQFQw3GvJaCGqMmikB48QNTwDTPQ`
I see there:

> Swarm listening on /ip4/127.0.0.1/tcp/45705/p2p/12D3KooWB4e2uekQUyxu1Fg37vrZwqi1XDSMt8Ss4X7DNuQvqxMz
Swarm listening on /ip4/192.168.178.120/tcp/45705/p2p/12D3KooWB4e2uekQUyxu1Fg37vrZwqi1XDSMt8Ss4X7DNuQvqxMz
Swarm listening on /ip4/172.18.0.1/tcp/45705/p2p/12D3KooWB4e2uekQUyxu1Fg37vrZwqi1XDSMt8Ss4X7DNuQvqxMz
num items in db:  0
Script is running. Press CTRL+C to terminate.
new entry 0:
 {
  v: 2,
  id: '/orbitdb/zdpuArrSWat42iJuw8LTqHQFQw3GvJaCGqMmikB48QNTwDTPQ',
  key: '03a3bc557f41c58850428c81dfd6f97b9e51498d792d10a9634496091992308543',
  sig: '304402206f70120f98c41818115c24f7a233e3611002275563d80dcc2f24b5d7ad4e3f63022029a71c9616f344034931351cf9b7286d5223c567bf307ff9e4c790625bb853ec',
  next: [ 'zdpuAstXaUGVDhwyZhq23iFZibTgGBfg5fW4tNkMfa4QJEkV1' ],
  refs: [
    'zdpuAsxfqF3JQ4tVAx9rfjSKpTbxSLTuvVoG1yH5M8HC23rDv',
    'zdpuAkx3G3LjcStnUZdSYoxjGu49osTEGLYxKV7zSWALhn4Y4',
    'zdpuAtFofBgZBkF91XiErL88ensRT8LdgbLh5JWKHWKHMKFfk',
    'zdpuB3ATanncdXJmC6r4fs6tpHQP9xnSGyxSJujrvNHrU9Z93',
    'zdpuArTkdYJYBHG6uj42aYNJui7L3MSijhTFeFvnxsB5H339Q',
    'zdpuAzdtUN4vqoYG3riDQmp6q48ANwni6us7SsANLyE7cHG34',
    'zdpuAxZYCXSKQQV4bNuqr84Eafs13Ghx6gs6SaB1Gzjs5vwVC',
    'zdpuApUu5aorqLLgscFJ7Lu94DynLpQRRkvbvMLZzC57tP9UV'
  ],
  clock: {
    id: '03a3bc557f41c58850428c81dfd6f97b9e51498d792d10a9634496091992308543',
    time: 10
  },
  payload: {
    op: 'PUT',
    key: 'e30630c1-fe17-46b2-91fd-d1a51e16fc89',
    value: {
      _id: 'e30630c1-fe17-46b2-91fd-d1a51e16fc89',
      content: 'content 9'
    }
  },
  identity: 'zdpuArR5VgUTnh3H6CVtKb4pKvgxiR1SX5SQP2tcTCvWZSZq2',
  hash: 'zdpuAtNf8rYbFDRfGLE3dVUdcmm229NzfHUK5PfJbGCCck6hp',
  bytes: Uint8Array(1025) [
    169,  97, 118,   2,  98, 105, 100, 120,  58,  47, 111, 114,
     98, 105, 116, 100,  98,  47, 122, 100, 112, 117,  65, 114,
    114,  83,  87,  97, 116,  52,  50, 105,  74, 117, 119,  56,
     76,  84, 113,  72,  81,  70,  81, 119,  51,  71, 118,  74,
     97,  67,  71, 113,  77, 109, 105, 107,  66,  52,  56,  81,
     78,  84, 119,  68,  84,  80,  81,  99, 107, 101, 121, 120,
     66,  48,  51,  97,  51,  98,  99,  53,  53,  55, 102,  52,
     49,  99,  53,  56,  56,  53,  48,  52,  50,  56,  99,  56,
     49, 100, 102, 100,
    ... 925 more items
  ]
}
peer joining db: 12D3KooWQLC5VRMTkgmzTo1Yb2RQahc8HQET8iTo1ZoULqDNQg32
file:///home/test/orbitdb/node_modules/@achingbrain/nat-port-mapper/dist/src/upnp/fetch.js:43
                throw new Error(`Request failed: ${response.statusCode}`); // eslint-disable-line @typescript-eslint/restrict-template-expressions
                ^
Error: Request failed: 500
    at ClientRequest.<anonymous> (file:///home/agw/orbitdb/node_modules/@achingbrain/nat-port-mapper/dist/src/upnp/fetch.js:43:23)
    at ClientRequest.emit (node:events:514:28)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:693:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
    at Socket.socketOnData (node:_http_client:535:22)
    at Socket.emit (node:events:514:28)
    at addChunk (node:internal/streams/readable:545:12)
    at readableAddChunkPushByteMode (node:internal/streams/readable:495:3)
    at Readable.push (node:internal/streams/readable:375:5)
    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)
Node.js v20.10.0

I get only the last result. ( content: 'content 9')
The exception at the end is new. I have never seen it before.

Here are the 2 clients I'm using.

orbit1.js:

~~~go
import { createOrbitDB } from '@orbitdb/core'
import { create } from 'ipfs-core'
import { v4 as uuidv4 } from 'uuid';
import {EventEmitter, setMaxListeners} from "events";
import IPFSBlockStorage from "@orbitdb/core/src/storage/ipfs-block.js";

const config1 = {
  Addresses: {
    API: '/ip4/127.0.0.1/tcp/0',
    Swarm: ['/ip4/0.0.0.0/tcp/0'],
    Gateway: '/ip4/0.0.0.0/tcp/0',
    NoAnnounce: [
    ]
  },
  Swarm: {
    AddrFilters: null,
  },
  Bootstrap: [],
  Discovery: {
    MDNS: {
      Enabled: true,
      Interval: 0
    }
  },
  Routing: {
    Routers: {
      Type: "dht",
      Parameters: {
        Mode: "auto",
        PublicIPNetwork: false,
        AcceleratedDHTClient: false
      }
    }
  }
}

EventEmitter.defaultMaxListeners = 20;

const ipfs1 = await create({ preload: { enabled: false }, offline: false, EXPERIMENTAL: { pubsub: true }, config: config1, repo: './ipfs/1' })

const orbitdb1 = await createOrbitDB({ ipfs: ipfs1, id: 'node1', directory: './orbitdb/1' })

const db1 = await orbitdb1.open('subscription-data', {type: 'documents'})
console.log("db1 address: ", db1.address)

for (let i=0; i<10; i++) {
  console.log("putting data", i)
  await db1.put({_id: uuidv4(), content: "content "+i})
}


async function handleTerminationSignal() {
  console.log('Received termination signal. Cleaning up and exiting...');
  await db1.close()
  await orbitdb1.stop()
  await ipfs1.stop()
  process.exit();
}

process.on('SIGTERM', handleTerminationSignal);
process.on('SIGINT', handleTerminationSignal);

console.log('Script is running. Press CTRL+C to terminate.');
~~~
orbit2.js:

~~~go
import { createOrbitDB } from '@orbitdb/core'
import { create } from 'ipfs-core'
import {EventEmitter, setMaxListeners} from "events";
import { createRequire } from "module";
const require = createRequire(import.meta.url);

const config2 = {
  Addresses: {
    API: '/ip4/127.0.0.1/tcp/0',
    Swarm: ['/ip4/0.0.0.0/tcp/0'],
    Gateway: '/ip4/0.0.0.0/tcp/0',
    NoAnnounce: [
    ]
  },
  Swarm: {
    AddrFilters: null,
  },
  Bootstrap: [],
  Discovery: {
    MDNS: {
      Enabled: true,
      Interval: 0
    }
  },
  Routing: {
    Routers: {
      Type: "dht",
      Parameters: {
        Mode: "auto",
        PublicIPNetwork: false,
        AcceleratedDHTClient: false
      }
    }
  }
}

EventEmitter.defaultMaxListeners = 20;

const argv = require('minimist')(process.argv.slice(2))
if (!argv.dbAddress) {
  console.error("dbAddress not passed")
  throw new Error("dbAddress not passed")
}

const ipfs2 = await create({ preload: { enabled: false }, offline: false, EXPERIMENTAL: { pubsub: true }, config: config2, repo: './ipfs/2' })
const orbitdb2 = await createOrbitDB({ ipfs: ipfs2, id: 'node2', directory: './orbitdb/2' })
const db2 = await orbitdb2.open(argv.dbAddress)

console.info("num items in db: ", (await db2.all()).length)

db2.events.on('join', async (peerId, heads) => {
  // The peerId of the ipfs1 node.
  console.log("peer joining db: " + peerId)
})

let updateCount=0
db2.events.on('update', async (entry) => {
  console.log(`new entry ${updateCount++}:\n`, entry)
})

async function handleTerminationSignal() {
  console.log('Received termination signal. Cleaning up and exiting...');
  await db2.close()
  await orbitdb2.stop()
  await ipfs2.stop()
  process.exit();
}

process.on('SIGTERM', handleTerminationSignal);
process.on('SIGINT', handleTerminationSignal);
console.log('Script is running. Press CTRL+C to terminate.');
~~~

**Comment by koh-osug** [comment]
When debugging it I can see that `receiveHeads is called:

~~~javascript
  const receiveHeads = (peerId) => async (source) => {
    for await (const value of source) {
      const headBytes = value.subarray()
      if (headBytes && onSynced) {
        await onSynced(headBytes)
      }
    }
    await onPeerJoined(peerId)
  }
~~~

and from this `applyOperation` is called, but only with the last synced item.

**Comment by haydenyoung** [comment]
At a glance, the problem you are experiencing could be related to https://discuss.ipfs.tech/t/importing-js-ipfs-gives-some-webpack-errors-reactjs/13037/6.

**Comment by koh-osug** [comment]
I have created a new issue with a test project related to this in #1148, since the initial topic was something else here.
